zaza/openstack/charm_tests/test_utils.py

```diff
@@ -133,6 +133,31 @@ def setUpClass(cls, application_name=None, model_alias=None):
             model_name=cls.model_name)
         logging.debug('Leader unit is {}'.format(cls.lead_unit))
 
+    def config_current(self, application_name=None, keys=None):
+        """Get Current Config of an application normalized into key-values.
+
+        :param application_name: String application name for use when called
+                                 by a charm under test other than the object's
+                                 application.
+        :type application_name:  Optional[str]
+        :param keys: iterable of strs to index into the current config.  If
+                     None, return all keys from the config
+        :type keys:  Optional[Iterable[str]]
+        :return: Dictionary of requested config from application
+        :rtype: Dict[str, str]
+        """
+        if not application_name:
+            application_name = self.application_name
+        _app_config = model.get_application_config(application_name)
+        # convert the more elaborate config structure from libjuju to key-value
+        keys = keys or _app_config.keys()
+        # note that conversion to string for all values is due to
+        # attempting to set any config with other types lead to Traceback
```

Comment:
s/lead/leads
Though the whole sentence could be clearer.

Beyond the comment, I am not sold on changing the types. I would be interested to see the traceback when you don't set things to a string.

Reply:
if i remember rightly, the traceback was trying to cast a go native type into a python type?  That may be why str(*) is used here.  Ill see if i can reproduce.  

Also, I think I remember that passing the config back with `model.set_application_config` fussed about the types?

Reply:
yeah... here's the issue.  You can 'get' the current config such that the types are preserved (aka -- not mapped through str) but when you play them back into the model.set_application_config -- it doesn't like anything that isn't str typed.  
```
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/zaza/model.py", line 531, in async_set_application_config
2020-01-24 12:40:44 [INFO]     .set_config(configuration))
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/juju/application.py", line 383, in set_config
2020-01-24 12:40:44 [INFO]     return await app_facade.Set(application=self.name, options=config)
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/juju/client/facade.py", line 471, in wrapper
2020-01-24 12:40:44 [INFO]     reply = await f(*args, **kwargs)
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/juju/client/_client8.py", line 1124, in Set
2020-01-24 12:40:44 [INFO]     reply = await self.rpc(msg)
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/juju/client/facade.py", line 607, in rpc
2020-01-24 12:40:44 [INFO]     result = await self.connection.rpc(msg, encoder=TypeEncoder)
2020-01-24 12:40:44 [INFO]   File "/home/ubuntu/charms/layers/keystone-ldap/build/builds/keystone-ldap/.tox/func-smoke/lib/python3.7/site-packages/juju/client/connection.py", line 456, in rpc
2020-01-24 12:40:44 [INFO]     raise errors.JujuAPIError(result)
2020-01-24 12:40:44 [INFO] juju.errors.JujuAPIError: json: cannot unmarshal bool into Go struct field ApplicationSet.options of type string
```